### PR TITLE
Use permanent errors in the taskrun reconciler

### LINKED ISSUE #2474: [feature] Start using knative PermanentErrors in reconcilers
# Expected Behavior
Distinguish explicitly between permanent errors and transient ones in the reconciler code.

# Actual Behavior
Today we capture errors that are considered permanent in `Reconcile` so that we can return `nil` and the key is not directly requeued. So we use the same semantic to indicate no error and permanent error.

[Knative pkg](https://github.com/knative/eventing/blob/9b0599f33985020f445820ce6ee843d5ab701122/vendor/knative.dev/pkg/controller/controller.go#L462-L466) defines `PermanentError`: when `Reconcile` is invoked on an object taken from the queue, [if it returns `PermanentError` the object is not requeued](https://github.com/knative/eventing/blob/9b0599f33985020f445820ce6ee843d5ab701122/vendor/knative.dev/pkg/controller/controller.go#L425). 
This allows being explicit in the reconcile code about which error is considered a permanent one.

/kind feature

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Instead of returning nil on error during reconcile, if the error
is transient return it, so that the key is requeue. If the error
is permanent return a permanent error, so that the key is not
requeued.

Fixes #2474

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

/kind cleanup

### CODE REVIEW

File: pkg/reconciler/taskrun/taskrun_test.go [hunk 1]
```diff
@@ -1326,6 +1326,17 @@ func TestReconcile_SetsStartTime(t *testing.T) {
...
+
+	t.Logf("Creating SA %s in %s", "default", "foo")
+	if _, err := clients.Kube.CoreV1().ServiceAccounts("foo").Create(&corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "default",
+			Namespace: "foo",
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
```
Comment (vdemeester) [inline pkg/reconciler/taskrun/taskrun_test.go:1339]: Why is this needed ? :thinking:

Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun_test.go:1339]: If you remove it the test will fail with missing service account.
It looks like returning the error exposes some hidden issues in existing tests.

Comment (vdemeester) [reply pkg/reconciler/taskrun/taskrun_test.go:1339]: ah ok :sweat_smile:

File: pkg/reconciler/taskrun/taskrun.go [hunk 2]
```diff
@@ -437,19 +438,21 @@ func (c *Reconciler) handlePodCreationError(ctx context.Context, tr *v1beta1.Tas
 			Reason:  podconvert.ReasonExceededResourceQuota,
 			Message: fmt.Sprintf("%s: %v", msg, err),
 		})
+		// return a transient error, so that the key is requeued
+		return err
```
Comment (zhujian7) [inline pkg/reconciler/taskrun/taskrun.go:442]: Hi, @afrittoli  AFAIK, the key will be requeued by the `SetTaskRunTimer` function, I am curiously about if the key will be requeued twice if we returned a transient error here?
